### PR TITLE
basic grid vs stack views

### PR DESCRIPTION
**Pull request by sofroniewn**
# Description
This PR follows up on #561 by implementing a basic grid vs stack view. Right now there are 3 methods that have gotten added to the viewer - `subplot(layer, position, size)` that will place a layer at a particular position in a grid of a particular size, and then two convenience methods `grid_view` and `stack_view` that will place layers in those respective configurations. `grid_view` also takes some parameters in case you want to set the grid to be a specific size or move through the layers with a particular stride, so that you could for example place layer 0, 1 on top of eachother in one grid cell, and layer 2, 3 on top of eachother in the next grid cell etc.

This is still WIP, I need to test it works in all configs and add some tests, but I wanted to quickly follow up on the discussion we were having in #561 with @jni @royerloic and @kellman

Here's a gif of the current performance:

grid_stack_views

It might be nice to add a button to the toggle grid vs stack view. Also should rearranging the order of the layers in the layers list rearrange the position of the layers in the grid once the grid has been made? If so in what way? Should there be away to rearrange the position of the layers in the grid just by clicking and dragging on them?

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by sofroniewn** [comment]
I'll also note that this implementation will work with all our different layer types

**Comment by royerloic** [comment]
Perfect, this is going to be super useful. Probably going to use it everyday...  As is anybody that needs to process images and compare them...

**Comment by sofroniewn** [comment]
This PR is now ready for review @jni @royerloic 

It is implemented as we discussed including a new button to toggle between grid and stacked mode. See here for its basic usage:

grid_mode

You can actually use the grid mode with nD data too, and with 3D rendering if you choose, see here:

nD_grid_mode

Finally if you want to use the stride parameter or set a custom grid size you have to use

```python
viewer.grid_view(n_row, n_column, stride)
viewer.stack_view()
```

**Comment by sofroniewn** [comment]
Seems like our CI is failing to even install windows right now

**Comment by jni** [comment]
@sofroniewn you are blowing my mind :heart_eyes:

**Comment by sofroniewn** [comment]
ah - and one other thing that I'm curious for @jni @royerloic opinion on is right now the layers list is ordered such that the most recently added layer goes on top and the oldest layer goes on the bottom. This is different than the natural ordering of the list, and Kira raised this issue #122 to discuss this a while back. At the time I liked the ordering with the new layers ontop, reasoning that those were the ones you were most recently adding were the ones you wanted to see the most. Now actually the layers list would automatically scroll if we grew it from the bottom, so you would see the most recently added at the bottom anyway.

I think it might make sense with grid mode to have the list ordered in the other way now with oldest layer at the top, and grow the layer going down. The reason for that is that the oldest layer gets added to the (0, 0) spot in the grid, and it's weird when the (0, 0) spot and the layer at the top of the grid don't match.

Thoughts?

**Comment by royerloic** [comment]
Wow, this is so cool. I love how the 3D grid just works...

**Comment by royerloic** [comment]
> ah - and one other thing that I'm curious for @jni @royerloic opinion on is right now the layers list is ordered such that the most recently added layer goes on top and the oldest layer goes on the bottom. This is different than the natural ordering of the list, and Kira raised this issue #122 to discuss this a while back. At the time I liked the ordering with the new layers ontop, reasoning that those were the ones you were most recently adding were the ones you wanted to see the most. Now actually the layers list would automatically scroll if we grew it from the bottom, so you would see the most recently added at the bottom anyway. 
> I think it might make sense with grid mode to have the list ordered in the other way now with oldest layer at the top, and grow the layer going down. The reason for that is that the oldest layer gets added to the (0, 0) spot in the grid, and it's weird when the (0, 0) spot and the layer at the top of the grid don't match.

Here we can think about the model/view separation to inform us in the best way.
I think that the order of the list should not change, that's the model, how we choose 
to display it is a different thing.  This could cause all sorts of trouble
where scripts would expect the last computed  image at the top, etc... think about 'stack' 
based computation for example. I am perfectly happy with having the top (first) layer
be always at (0,0). The grid would be reorganised. Maybe that's confusing. 

Oh wait, you might be suggesting simply to display the other way around, not to
reorder the list itself... Yes. if that's what you mean, then having the top of the list 
displayed at the end of the grid. It might be confusing though... If you just display
two layers, it is not obvious which layer corresponds to what... One possibility is to
display a 'border' (highlight) that is synced with the layer list selection. I am not sure
that we want to go this path now, though.

Not obvious what is best. We need to think about it more.

**Comment by jni** [comment]
I don't really get the confusion... If I remember correctly, we *append* to the layer list (?), and the list grows towards the top? This is actually fine and if you think about a 2D images + a right-handed reference frame, this is the direction you expect to grow in when you have (row, col) coordinates.

Now, if you take the linear list and reshape it using `np.reshape`, then you end up with the list growing on the bottom-right edge. So that remains consistent as far as I can tell...

So, my vote:

- layer list grows by appending
- latest layer on top in the layer list
- latest layer on the bottom-right in the grid view

I do see that most users would read left-right then top-bottom, but I don't think that is our primary concern...

**Comment by sofroniewn** [comment]
Yeah @royerloic i’m not thinking of changing the model, which would be very complex, just the view.

Should new layers appear at the bottom or top of the view. Right now it is the top, maybe it should be the bottom? That’s all I am asking

**Comment by royerloic** [comment]
Top seems reasonable to me, new layers get added at the top of the list,  and at the left-top corner of the grid. I am afraid we  have a tie with @jni . @sofroniewn feel free to pick your preference if Juan agrees :-)

**Comment by sofroniewn** [comment]
Hmm ok - I think I will consult @liaprins-czi for a 3rd opinion on monday. I think we're all in agreement that the layers list gets appended too though as otherwise all the indexing is weird.

As to growing the layers list view at the top or bottom, and growing the grid view at the top left / bottom right. I think I actually lean towards grow the list view at the bottom and grow the grid view at the bottom right.

For growing the list view at the bottom it is then much easier to count down from the top and know that the top layer is the 0th element in the list, the next layer is the 1st etc, whereas now you need to count up from the bottom. Purely from a code standpoint now I have to do all sorts of weird reversals to maintain the correspondence between the layer list and its view, which is fine as they happen under the hood, but maybe an indication something isn't natural either.

The rest of this PR is ready for review

**Comment by royerloic** [comment]
ok, makes sense, everything grows at the bottom (right). In that case it is all very intuitive and consistent. It was late yesterday... ;-)

**Comment by sofroniewn** [comment]
What does @jni think? I'll add with this screenshot, in the current view it just looks confusing that the two lists run in opposite orders:

layers_order

**Comment by jni** [comment]
I'm pretty happy to think of the layerlist view as a "stack", because the images do appear like they are stacked together and the topmost image is hiding all the other ones. Then, think about putting your palm on the topmost image and sliding it out, splaying all the "pictures" underneath across the "desk" towards the bottom right. (In fact this could be a neat animation to add. ;) It is that one metaphor that breaks down if you invert the order of the layerlist. But otherwise, your arguments make sense, and I definitely prefer inverting the list to inverting the order of the montage.

I'm very curious what @liaprins-czi will say. =)

File: examples/layers.py
Comment (jni) [inline examples/layers.py:19]: Do you want to set the layout mode to grid here?

**Review by jni — APPROVED** [review]
@sofroniewn the implementation is wonderfully simple! And the current layout matches my preference, so feel free to merge immediately! ;)

**Comment by sofroniewn** [comment]
During discussion with @liaprins-czi today, her preference was for the layers in the layers list view to remain as they are with most visible on top, and therefore most recent on top, as you want the most recently added layer also to be most visible. Given that, she felt it was natural to have the most recent layer in the top left corner (which is different then our current implementation where it is bottom right). She felt it was important to have that consistency between the top left corner of the grid view and the top of the layers list.

The above implementation still leaves us with a mismatch between the order of the layers list view and the order of the layers in the actual python list, but she felt this was ok - it was more important to have consistency within the GUI then between how things are done within the GUI and within the code, which I agree with, it isn't that big a deal that the ordering is different, and we should really be using the layer name to index into the layers list rather than the location in the list.

Finally she thought it was strange and confusing that the layers could be edited in grid mode, and that you could enter 3D rendering in grid mode, and that you could create new layers in grid mode. All of these options actually led us to weird confusing non-intuitive behavior and she proposed to disable as much of this functionality (at least from the GUI) when in grid mode so as users don't click the wrong stuff and end up doing something weird. I'll think about the best way to implement this, but broadly agree.

If this sounds good to @jni and @royerloic I'll probably make these changes tomorrow

**Comment by jni** [comment]
@sofroniewn 

> She felt it was important to have that consistency between the top left corner of the grid view and the top of the layers list.

*sigh* fine. =P Actually I'd very much like it if we could make this a configurable option somewhere, so I could play around with both modes. I have a strong intuition that having the next added image on the bottom right would work better, at least for me, someone experienced in the NumPy ecosystem. For biologists who haven't necessarily worked with arrays, I'll admit it might be a different story...

> she thought it was strange and confusing that the layers could be edited in grid mode, and that you could enter 3D rendering in grid mode, and that you could create new layers in grid mode.

Here I'm actually in favour of keeping the current behaviour. One common use case that would be enabled basically as soon as this PR is merged is to pick correspondences between two images you want to co-register. This would actually be quite easy to do with the two images, two interleaved points layers, and a stride of 2.

And both @royerloic and I were very impressed with the side-by-side volume rendering, and I think it would be a very nice way to see certain datasets.

**Comment by sofroniewn** [comment]
@jni ok your points are well taken - i wont disable anything now, we can always disable for the `0.2.1` release, but merging the PR now will give people a change to play around with it.

I've made the default order in the grid such that the most visible layer goes in the top left corner, but if you pass a negative stride (say `stride=-1`) then the order will be reversed and it will go in the bottom right corner. This seems like a pretty reasonable compromise to me.

I've also added tests. I will merge once they all pass.

File: examples/layers.py
Comment (sofroniewn) [reply examples/layers.py:19]: done

**Pull request merged**
